include/zephyr/llext/elf.h

```diff
@@ -0,0 +1,396 @@
+/*
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+#ifndef ZEPHYR_LLEXT_ELF_H
+#define ZEPHYR_LLEXT_ELF_H
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief ELF types and parsing
+ * @defgroup elf ELF data types and defines
+ * @ingroup llext
+ * @{
+ */
+
+/**
+ * @brief Type aliases for 32 bit ELF
+ * @defgroup elf_32bit_types ELF 32bit typedefs
+ * @{
+ */
+
+typedef uint32_t elf32_addr;
+typedef uint16_t elf32_half;
+typedef uint32_t elf32_off;
+typedef int32_t elf32_sword;
+typedef uint32_t elf32_word;
+
+/**
+ * @}
+ */
+
+/**
+ * @brief Type aliases for 64 bit ELF
+ * @defgroup elf_64bit_types ELF 62bit typedefs
+ * @{
+ */
+
+typedef uint64_t elf64_addr;
+typedef uint16_t elf64_half;
+typedef uint64_t elf64_off;
+typedef int32_t elf64_sword;
+typedef uint32_t elf64_word;
+typedef int64_t elf64_sxword;
+typedef uint64_t elf64_xword;
+
+/**
+ * @}
+ */
+
+
+/**
+ * @brief ELF identifier block
+ *
+ * 4 byte magic (.ELF)
+ * 1 byte class (Invalid, 32 bit, 64 bit)
+ * 1 byte endianness (Invalid, LSB, MSB)
+ * 1 byte version (1)
+ * 1 byte OS ABI (0 None, 1 HP-UX, 2 NetBSD, 3 Linux)
+ * 1 byte ABI (0)
+ * 7 bytes padding
+ */
+#define EI_NIDENT 16
+
+/**
+ * @brief ELF Header(32-bit)
+ */
+struct elf32_ehdr {
+	unsigned char e_ident[EI_NIDENT];
+	elf32_half e_type;
+	elf32_half e_machine;
+	elf32_word e_version;
+	elf32_addr e_entry;
+	elf32_off e_phoff;
+	elf32_off e_shoff;
+	elf32_word e_flags;
+	elf32_half e_ehsize;
+	elf32_half e_phentsize;
+	elf32_half e_phnum;
+	elf32_half e_shentsize;
+	elf32_half e_shnum;
+	elf32_half e_shstrndx;
+};
+
+/**
+ * @brief ELF Header(64-bit)
+ */
+struct elf64_ehdr {
+	unsigned char e_ident[EI_NIDENT];
+	elf64_half e_type;
+	elf64_half e_machine;
+	elf64_word e_version;
+	elf64_addr e_entry;
+	elf64_off e_phoff;
+	elf64_off e_shoff;
+	elf64_word e_flags;
+	elf64_half e_ehsize;
+	elf64_half e_phentsize;
+	elf64_half e_phnum;
+	elf64_half e_shentsize;
+	elf64_half e_shnum;
+	elf64_half e_shstrndx;
+};
+
+/**
+ * @brief ELF file type
+ * @defgroup elf_file_type ELF file types
+ * @ingroup elf
+ * @{
+ */
+
+/** Relocatable (unlinked) ELF */
+#define ET_REL  1
+
+/** Executable (without PIC/PIE) ELF */
+#define ET_EXEC 2
+
+/** Dynamic (executable with PIC/PIE or shared lib) ELF */
+#define ET_DYN  3
+
+/** Core Dump */
+#define ET_CORE 4
+
+/**
+ * @}
+ */
+
+/*
+ * Section Header(32-bit)
+ */
+struct elf32_shdr {
+	elf32_word sh_name;
+	elf32_word sh_type;
+	elf32_word sh_flags;
+	elf32_addr sh_addr;
+	elf32_off sh_offset;
+	elf32_word sh_size;
+	elf32_word sh_link;
+	elf32_word sh_info;
+	elf32_word sh_addralign;
+	elf32_word sh_entsize;
```

Comment:
would be nice to have docstrings for the fields (applies to a few other structs in this file)